p2p/sharding.py

```diff
@@ -0,0 +1,209 @@
+from p2p.protocol import (
+    Command,
+    Protocol,
+)
+from p2p.peer import (
+    BasePeer,
+    PeerPool,
+    PeerPoolSubscriber,
+)
+from p2p.p2p_proto import (
+    DisconnectReason,
+)
+from p2p.exceptions import (
+    HandshakeFailure,
+)
+
+
+COLLATION_PERIOD = 1
+
+
+class Status(Command):
+    _cmd_id = 0
+
+
+class Collations(Command):
+    _cmd_id = 1
+
+    structure = rlp.sedes.CountableList(Collation)
+
+
+class ShardingProtocol(Protocol):
+    name = "sha"
+    version = 0
+    _commands = [Status, Collations]
+    cmd_length = 2
+
+    logger = logging.getLogger("p2p.sharding.ShardingProtocol")
+
+    def send_handshake(self) -> None:
+        cmd = Status(self.cmd_id_offset)
+        self.logger.debug("Sending status msg")
+        self.send(*cmd.encode([]))
+
+    def send_collations(self, collations: List[Collation]) -> None:
+        cmd = Collations(self.cmd_id_offset)
+        self.logger.debug("Sending {} collations".format(len(collations)))
+        self.send(*cmd.encode(collations))
+
+
+class ShardingPeer(BasePeer):
+    _supported_sub_protocols = [ShardingProtocol]
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.incoming_collation_queue = None
+        self.known_collation_hashes = set()
+
+    def set_incoming_collation_queue(self, queue: asyncio.Queue) -> None:
+        self.incoming_collation_queue = queue
+
+    #
+    # Handshake
+    #
+    async def send_sub_proto_handshake(self) -> None:
+        self.sub_proto.send_handshake()
+
+    async def process_sub_proto_handshake(
+        self,
+        cmd: Command,
+        msg: protocol._DecodedMsgType
+    ) -> None:
+        if not isinstance(cmd, Status):
+            self.disconnect(DisconnectReason.other)
+            raise HandshakeFailure("Expected status msg, got {}, disconnecting".format(cmd))
+
+    #
+    # Receiving Collations
+    #
+    def handle_sub_proto_msg(self, cmd: Command, msg: protocol._DecodedMsgType) -> None:
+        if isinstance(cmd, Collations):
+            self._handle_collations_msg(msg)
+        else:
+            super().handle_sub_proto_msg(cmd, msg)
+
+    def _handle_collations_msg(self, msg: List[Collation]) -> None:
+        for collation in msg:
+            self.known_collation_hashes.add(collation.hash)
+            if self.incoming_collation_queue is not None:
+                self.incoming_collation_queue.put_nowait(collation)
+
+    def send_collations(self, collations: List[Collation]) -> None:
+        for collation in collations:
+            if collation.hash not in self.known_collation_hashes:
+                self.known_collation_hashes.add(collation.hash)
+                self.sub_proto.send_collations(collations)
+
+
+class ShardSyncer(PeerPoolSubscriber):
+
+    def __init__(
+        self,
+        shard: Shard,
+        mean_proposing_period: int,
+        peer_pool: PeerPool,
+        token: CancelToken = None,
+    ) -> None:
```

Comment:
Same beef with this indentation style.  Hard to visually parse imho.  Consider either:

```python
def my_method(self
              arg_2,
              arg_3) -> None:
# or
def my_method(
        self,
        arg_2,
        arg_3) -> None:
    thing = ...
```